### DISCUSSION TITLE
Default dateFormat in PrimeVue not using locale's format

### DISCUSSION THREAD

**Discussion by Petri-Oosthuizen** [primefaces/community#478] [PrimeVue] [answered]
Currently, the default dateFormat behaviour does not consider the locale's date format. Instead, it defaults to the "mm/dd/yyyy" format commonly used in the United States. This may lead to confusion for users outside the U.S. who expect the default date format to align with their locale.

As it currently stands it is up to each dev to manually figure out what the `dateFormat` of a specific locale is.

```
const app = createApp(App);

app.use(PrimeVue, {
  ripple: true,
  locale: {
    dateFormat: "dd/mm/yy", // <----- This should be automatically determined based on the locale
  },
});

app.mount("#app");
```

I am open to suggestions as I may not have a complete understanding of the expected implementation.

**Comment by melloware** [comment, marked as answer]
But can't you use [PrimeLocale](https://github.com/primefaces/primelocale) and just import the locale for your user?  

https://primevue.org/configuration/#locale

**Reply by Petri-Oosthuizen** [reply]
I'll give it a go thanks!

**Reply by melloware** [reply]
You can even detect the current browser language and switch to the correct PrimeLocale file.

**Reply by Petri-Oosthuizen** [reply]
The readme does not have a lot of details on the usage of the files. Firstly, how are they intended to be used? Is the intended approach to reference the git repo in a project's package.json? Secondly, are there perhaps any utility functions to get the correct PrimeLocale file for the current (or a specific) locale or default to "en"? I can imagine that would be quite useful for everyone using this package.

**Reply by melloware** [reply]
See this for how to import the files in your package.json: https://github.com/primefaces/primelocale/issues/71#issuecomment-1662424247

Its not published to NPM yet but its on PrimeTek's radar.

I am not an expert in Vue but my guess is like this example in onMounted you would do something like this psuedocode:

```js
export default defineComponent({
    setup() {
        onMounted(() => {
            const browserLocale = (navigator.languages || [])[0] || navigator.userLanguage || navigator.language || navigator.browserLanguage || 'en';
            const primevue = usePrimeVue();
            primevue.config.locale = {...primelocales[browserLocale]}
        })
    }
});
```

once again the above is psuedo-code just to give you the idea how to do it.

**Reply by Petri-Oosthuizen** [reply]
Thanks! Small typo in that comment, it should be `primefaces` not `primfaces`.

**Reply by melloware** [reply]
Yep sorry was just slinging a guess!

**Reply by Petri-Oosthuizen** [reply]
Out of curiosity, how did you import primelocales?

**Reply by melloware** [reply]
you know what let me put a Code Sandbox together and make sure this works the way I expect.

**Reply by melloware** [reply]
Its using React but you get the idea: https://stackblitz.com/edit/react-3glwrd?file=src%2FApp.tsx,package.json,src%2Findex.tsx

**Reply by Petri-Oosthuizen** [reply]
Awesome thanks, I see you are [importing the files individually](https://stackblitz.com/edit/react-3glwrd?file=src%2FApp.tsx%3AL12,package.json,src%2Findex.tsx) and not dynamically.
```
import ru from 'primelocale/ru.json';
import tr from 'primelocale/tr.json';
```

I was thinking of doing something like ```const localeFile = await import(`primelocale/${locale}.json`);``` but I've not worked through the kinks yet (this does not handle locales for which no files exist for eg.)

Interesting that Russian on my computer displays the date as `16.11.20232023`:
<img width="363" alt="image" src="https://github.com/primefaces/community/assets/4954611/7ac68d90-0924-404b-8f74-2244f411eed5">

**Reply by melloware** [reply]
Yep just wanted to give you the idea but you can absolutely dynamically import them.

**Reply by Petri-Oosthuizen** [reply]
I'm leaving the following here as I had some issues with locales.

If you come across locale discrepancies, keep in mind that `navigator.language` may not match `new Intl.DateTimeFormat().resolvedOptions().locale`. The reason is that the former relies on browser language preferences, while the latter relies on the operating system's language preferences.

In my particular scenario, I observed a discrepancy where `navigator.language` returned "en-US," while `new Intl.DateTimeFormat().resolvedOptions().locale` returned "en-GB." This resulted in inconsistencies, causing certain components to format my date as mm/dd/yy and others as dd/mm/yy.

**Reply by melloware** [reply]
I have seen that same scenario for a UK client. Here is what I did that was kinda lame.

```js
/**
   * Grab the user's timezone from the browser
   */
private static readonly TIMEZONE = Intl.DateTimeFormat().resolvedOptions().timeZone;

const isUsEnglish = timezone.match('America/');
```

**Reply by Petri-Oosthuizen** [reply]
In response to our earlier discussion, I've quickly (and dirtily) implemented dynamic importing of locale files using Vite with the following approach:

```js
// Get the user locale-lang preference from the browser
const browserLocale = getBrowserLocale();

// Use the OS locale as a fallback
const osLocale = new Intl.DateTimeFormat().resolvedOptions().locale;

// Load all primevue locale files, the relative path is ugly but `@/` or `./` did not work.
const locales = import.meta.glob("../../../node_modules/primelocale/*.json");

const localeRe = new RegExp(`\\b${browserLocale}.json$`, "i");

let localeFileName = Object.keys(locales).find((locale) => localeRe.test(locale));

// Additional checks for if a file was not found for the locale are omitted for brevity.
// Here I check the operating system locale, browser lang, and os lang before defaulting to `defaultOptions`.
// ...

const localeFile = Object.values(
  ((await locales[localeFileName]?.()) as any)?.default,
)[0] as PrimeVueLocaleOptions ?? defaultOptions.locale;

primevue.config.locale = localeFile;
```

**Reply by melloware** [reply]
Slick. You should create a StackBlitz demoing that so others can reference in the future!

**Reply by Petri-Oosthuizen** [reply]
Here's a very basic example. I'm open to improvements.

https://codesandbox.io/p/devbox/primevue-locale-loader-gxm2ct

https://github.com/primefaces/community/assets/4954611/1eda82f8-c416-4e6f-bf05-efb03f2422d7

**Reply by melloware** [reply]
This is fantastic

**Reply by melloware** [reply]
@Petri-Oosthuizen when I click your link above I get the default PrimeVue template can you update the link to your real showcase?

**Reply by Petri-Oosthuizen** [reply]
That is strange... I'll update it when I'm back in office on the 4th of January.

**Reply by Petri-Oosthuizen** [reply]
Please try the [updated link](https://codesandbox.io/p/devbox/primevue-locale-loader-gxm2ct)

**Reply by melloware** [reply]
Fantastic!

**Reply by BenJackGill** [reply]
I struggled for a long time with this. I wanted a Nuxt solution that would load the users default locale. Here is my version, adopted from the example provided above...

1) Create a composable to fetch the users locale. For this I created a composable [as shown by Nuxt](https://nuxt.com/docs/examples/advanced/locale) for detecting locale.

**composables/useLocale.ts**
```TypeScript
const getDefaultLocale = (fallback = "en-US") => {
  const locale = ref(fallback);
  if (import.meta.server) {
    const reqLocale = useRequestHeaders()["accept-language"]?.split(",")[0];
    if (reqLocale) {
      locale.value = reqLocale;
    }
  } else if (import.meta.client) {
    const navLang = navigator.language;
    if (navLang) {
      locale.value = navLang;
    }
  }
  return locale;
};

export const useLocale = () => {
  const defaultLocale = getDefaultLocale();
  return useState<string>("locale", () => defaultLocale.value);
};
```

2) Install the latest version of https://github.com/primefaces/primelocale using `npm add github:primefaces/primelocale`. It will look like this:

**package.json**
```JSON
  "dependencies": {
    "primelocale": "github:primefaces/primelocale",
  },
```

3) Create a composable to auto load the correct locale into `primeVue.config.locale`.

Note: I first tried creating this as a Nuxt plugin. But that was causes build failures. Use a composable instead, and call from app.vue.

**composables/usePrimeVueLocales.ts**
```TypeScript
import { type PrimeVueLocaleOptions, defaultOptions } from "primevue/config";

type LocaleModule = { default: Record<string, PrimeVueLocaleOptions> };

export const usePrimeVueLocales = async () => {
  const primeVue = usePrimeVue();

  // Get the user's locale (both server and client side)
  const localeStore = useLocaleStore();
  const { locale } = storeToRefs(localeStore);

  // Load all primevue locale files, the relative path is ugly but `@/` or `./` did not work.
  const primeVueLocaleFiles = import.meta.glob(
    "../node_modules/primelocale/*.json",
  );

  // First we try to find the locale (e.g. `en-US.json`)
  // If that fails we try to find a matching language (e.g. `en.json`)
  const localeRegex = new RegExp(`\\b${locale.value}.json$`, "i");
  const langRegex = new RegExp(`\\b${locale.value.split("-")[0]}.json$`, "i");

  const localeFileName =
    Object.keys(primeVueLocaleFiles).find((file) => localeRegex.test(file)) ||
    Object.keys(primeVueLocaleFiles).find((file) => langRegex.test(file));

  let localeFile = defaultOptions.locale;

  if (localeFileName) {
    const localeModule = (await primeVueLocaleFiles[
      localeFileName
    ]?.()) as LocaleModule;
    localeFile =
      (Object.values(localeModule?.default)[0] as PrimeVueLocaleOptions) ??
      defaultOptions.locale;
  }

  console.log("Loaded locale file:", localeFile);

  primeVue.config.locale = localeFile;
};
```

4) Call the composable from `app.vue`

**app.vue**
```Vue
<template>
  <NuxtLayout>
    <NuxtPage />
  </NuxtLayout>
</template>

<script setup lang="ts">
usePrimeVueLocales();
</script>
```

Extra reading:
- https://codesandbox.io/p/devbox/primevue-locale-loader-gxm2ct?file=%2Fsrc%2FApp.vue%3A28%2C38
- https://nuxt.com/docs/examples/advanced/locale
